subprojects/testing-native/src/main/java/org/gradle/nativeplatform/test/xctest/plugins/XCTestConventionPlugin.java

```diff
@@ -0,0 +1,164 @@
+import org.gradle.api.Plugin;
+import org.gradle.api.Task;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.file.Directory;
+import org.gradle.api.file.DirectoryVar;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.internal.file.FileOperations;
+import org.gradle.api.internal.project.ProjectInternal;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.specs.Spec;
+import org.gradle.api.tasks.TaskContainer;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.language.cpp.plugins.CppBasePlugin;
+import org.gradle.language.swift.internal.DefaultSwiftComponent;
+import org.gradle.language.swift.model.SwiftComponent;
+import org.gradle.language.swift.plugins.SwiftBasePlugin;
+import org.gradle.language.swift.tasks.SwiftCompile;
+import org.gradle.nativeplatform.platform.internal.DefaultNativePlatform;
+import org.gradle.nativeplatform.tasks.LinkExecutable;
+import org.gradle.nativeplatform.test.xctest.tasks.CreateXcTestBundle;
+import org.gradle.nativeplatform.test.xctest.tasks.XcTest;
+import org.gradle.nativeplatform.toolchain.NativeToolChain;
+import org.gradle.nativeplatform.toolchain.internal.NativeToolChainInternal;
+import org.gradle.nativeplatform.toolchain.internal.NativeToolChainRegistryInternal;
+import org.gradle.nativeplatform.toolchain.internal.PlatformToolProvider;
+import org.gradle.process.internal.ExecAction;
+import org.gradle.process.internal.ExecActionFactory;
+
+import javax.inject.Inject;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.util.Collections;
+
+/**
+ * A plugin that sets up the infrastructure for testing native binaries with XCTest test framework. It also adds conventions on top of it.
+ *
+ *
+ */
+@Incubating
+public class XCTestConventionPlugin implements Plugin<ProjectInternal> {
+    private final FileOperations fileOperations;
+    private final ExecActionFactory execActionFactory;
+
+    private static File cachedSdkPlatformDir;
+
+    @Inject
+    public XCTestConventionPlugin(FileOperations fileOperations, ExecActionFactory execActionFactory) {
+        this.fileOperations = fileOperations;
+        this.execActionFactory = execActionFactory;
+    }
+
+    @Override
+    public void apply(final ProjectInternal project) {
+        project.getPluginManager().apply(SwiftBasePlugin.class);
+
+        final DirectoryVar buildDirectory = project.getLayout().getBuildDirectory();
+        Directory projectDirectory = project.getLayout().getProjectDirectory();
+        ConfigurationContainer configurations = project.getConfigurations();
+        TaskContainer tasks = project.getTasks();
+
+        // Add the component extension
+        SwiftComponent component = project.getExtensions().create(SwiftComponent.class, "xctest", DefaultSwiftComponent.class, fileOperations);
+        component.getSource().from(projectDirectory.dir("src/test/swift"));
```

Comment:
If we're adding an extension, we'll need some test coverage to show that you can configure the test source locations. Can add this as a later PR.